logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -17,40 +17,99 @@
 
 module LogStash
   class PipelineState
-    attr_reader :pipeline_id, :pipeline
+    attr_reader :pipeline_id
 
     def initialize(pipeline_id, pipeline)
       @pipeline_id = pipeline_id
       @pipeline = pipeline
-      @reloading = Concurrent::AtomicBoolean.new(false)
+      @loading = Concurrent::AtomicBoolean.new(false)
+
+      # this class uses a lock to ensure thread safe visibility.
+      @lock = Mutex.new
     end
 
     def terminated?
-      # a reloading pipeline is never considered terminated
-      @reloading.false? && @pipeline.finished_execution?
+      @lock.synchronize do
+        # a loading pipeline is never considered terminated
+        @loading.false? && @pipeline.finished_execution?
+      end
     end
 
-    def set_reloading(is_reloading)
-      @reloading.value = is_reloading
+    def set_loading(is_loading)
+      @lock.synchronize do
+        @loading.value = is_loading
+      end
     end
 
     def set_pipeline(pipeline)
-      raise(ArgumentError, "invalid nil pipeline") if pipeline.nil?
-      @pipeline = pipeline
+      @lock.synchronize do
+        raise(ArgumentError, "invalid nil pipeline") if pipeline.nil?
+        @pipeline = pipeline
+      end
+    end
+
+    def pipeline
+      @lock.synchronize { @pipeline }
     end
   end
 
+  class PipelineStates
+
+    def initialize
+      @states = {}
+      @locks = {}
+      @lock = Mutex.new
+    end
+
+    def get(pipeline_id)
+      @lock.synchronize do
+        @states[pipeline_id]
+      end
+    end
+
+    def put(pipeline_id, state)
+      @lock.synchronize do
+        @states[pipeline_id] = state
+      end
+    end
+
+    def remove(pipeline_id)
+      @lock.synchronize do
+        @states.delete(pipeline_id)
```

Comment:
do we also need to dereference the `@locks[pipeline_id]`, or is the unbounded growth of per-pipeline lock cache safe enough because how many distinct pipelines can a process run anyway? If we do clean up here, we also have to mark the existing lock object as invalid, which is whole new can of worms.

Reply:
Nevermind my previous comment - we should definitely also delete from `@locks` ... this was an oversight.  Fixed.

Reply:
Now that `PipelineStates#remove(pipeline_id)` deletes locks, _unless we have some other mutual exclusion*_, it is possible that the lock we acquire with `PipelineStates#get_lock` (e.g., at the top of `PipelinesRegistry#create_pipeline`, `PipelinesRegistry#terminate_pipeline`, `PipelinesRegistry#reload_pipeline`, etc.) can be deleted by another thread without the lock-holder's knowledge, and a new lock can be created in its place.

*: I _believe_ that `PipelinesRegistry#create_pipeline` and friends ares only called when the agent's `@convergence_lock` is exclusively held (by way of the various `PipelineAction`s), but there are a number of layers of indirection and I haven't been able to pin it down to prove it. Even if they do though, we should find a way to make this more explicit lest new logic be introduced that accidentally circumvents the mutual exclusion we currently enjoy.

An alternative would be to never actually give a reference the lock out, instead implementing `PipelineStates#with_lock(pipeline_id)` that yields to a provided block after acquiring the lock.